File: programmers/코딩테스트_고득점kit/6_탐욕법.큰수만들기/junilhwang-01.js
// https://programmers.co.kr/learn/courses/30/lessons/42883?language=javascript

function solution (nums, k) {
  const arr = Array.from(nums)
  const last = nums.length, len = last - k
  let left = 0, idx = 0, num = ""
  for (let i = 0; i < len; i++) {
    let max = nums[left]
    while (left <= k + i) {
      const n = nums[left]
      if (max < n) max = n, idx = left
      let nextLeft = -1, nextMax = 9
      while (nextMax > ~~max && nextLeft === -1) nextLeft = nums.indexOf(nextMax--, left+1)
      if (nextLeft === -1) break
      left = nextLeft < k + 1 ? nextLeft : left + 1
    }
    num += max, left = ++idx
  }
  return num
}

console.log(solution("1924", 2), "94")
console.log(solution("1231234", 3), "3234")
console.log(solution("4177252841", 4), "775841")

File: programmers/난이도별/level02.더_맵게/junilhwang-01.py
# https://programmers.co.kr/learn/courses/30/lessons/42626
# 섞은 음식의 스코빌 지수 = 가장 맵지 않은 음식의 스코빌 지수 + (두 번째로 맵지 않은 음식의 스코빌 지수 * 2)
def solution(s, K):
    h = []
    for v in s : insert(h, v)
    if h[0] >= K : return 0
    for i in range(1, len(h)) :
        insert(h, delete(h) + delete(h) * 2)
        if h[0] >= K : return i
    return -1

def insert (h, n) :
    i = len(h)
    h.append(n)
    while i > 0 and n < h[int(i/2)] : h[i], i = h[int(i/2)], int(i/2)
    h[i] = n
    
def delete (h) :
    n, h[0] = h[0], h[-1]
    hlen = len(h)
    i = 0
    while i * 2 + 1 < hlen :
        next = i * 2 if i != 0 else 1
        if next + 1 < hlen and h[next] > h[next + 1] : next += 1
        h[next], h[i], i = h[i], h[next], next
    del h[-1]
    return n

print(solution([1, 2, 3, 9, 10, 12], 100))
print(solution([1, 2], 7))
print(solution([1], 2))
print(solution([5], 5))
print(solution([1,2,7], 5))
print(solution([2,3,7], 7))
print(solution([0], 0))